`AlgebraOfGraphics.jl` Chapter (#289)
JuliaDataScience/JuliaDataScience: contents/data_vis_aog_stats.md

## Statistical Visualizations {#sec:aog_stats}

`AlgebraOfGraphics.jl` can perform **statistical transformations** as layers with five functions:

- **`expectation`**: calculates the mean (_expectation_) of the underlying Y-axis column
- **`frequency`**: computes the frequency (_raw count_) of the underlying X-axis column
- **`density`**: computes the density (_distribution_) of the underlying X-axis column
- **`linear`**: computes a linear trend relationship between the underlying X- and Y-axis columns
- **`smooth`**: computes a smooth relationship between the underlying X- and Y-axis columns

Let's first cover `expectation`:

```jl
s = """
    # aog_expectation # hide
    CairoMakie.activate!() # hide
    plt = data(df) *
        mapping(:name, :grade) *
        expectation()
    f = draw(plt) # hide
    draw(plt)
    label = "aog_expectation" # hide
    caption = "AlgebraOfGraphics bar plot with expectation." # hide
    link_attributes = "width=60%" # hide
    Options(f.figure; filename=label, caption, label, link_attributes) # hide
    """
sco(s)
```

> As expected, `expectation` adds a statistical transformation layer that tells `AlgebraOfGraphics.jl` to compute the mean of the Y-axis values for every unique X-axis values.

Reviewer: ```suggestion
Here, `expectation` adds a statistical transformation layer that tells `AlgebraOfGraphics.jl` to compute the mean of the Y-axis values for every unique X-axis values.
```

In our case, it computed the mean of grades for every student.
Note that we could safely remove the visual transformation layer (`visual(BarPlot)`) since it is the default visual transformation for `expectation`.

> Following up, we'll show an example with `frequency`:

Reviewer: ```suggestion
Next, we'll show an example with `frequency`:
```

synthetic_df = DataFrame(; x, y)
first(synthetic_df, 5)
"""; process=without_caption_label
)
```

Let's begin with `linear`:

```jl
s = """
    # aog_linear # hide
    CairoMakie.activate!() # hide
    plt = data(synthetic_df) *
        mapping(:x, :y) *
        (visual(Scatter) + linear())
    f = draw(plt) # hide
    draw(plt)
    label = "aog_linear" # hide
    caption = "AlgebraOfGraphics scatter plot with linear trend estimation." # hide
    link_attributes = "width=60%" # hide
    Options(f.figure; filename=label, caption, label, link_attributes) # hide
    """
sco(s)
```

We are using the **distribute property** (@sec:aog) for more efficient code inside our `mapping`,
`a  * (b + c) = (a * b) + (a + b)`,
where:

- `a`: the `data` and `mapping` layers fused into a single layer
- `b`: the `visual` transformation layer
- `c`: the statistical `linear` transformation layer

`linear` adds a linear trend between the X- and Y-axis mappings with a 95% confidence interval shaded region.

Finally, the same example as before but now replacing `linear` with `smooth`:

```jl
s = """
    # aog_smooth # hide
    CairoMakie.activate!() # hide
    plt = data(synthetic_df) *

>         mapping(:x, :y) *

Reviewer 2: ```suggestion
        mapping(:x, :y) *
```
To be fully consistent with earlier code, this should be written as
```
       mapping(:x,
            :y
       )
```
